### DISCUSSION TITLE
Composite keys recommended?

### DISCUSSION THREAD

**Discussion by pke** [vercel/swr#1920] [Q&A] [answered]
I wonder if composite keys are actually recommended or should mutation happen via useEffect?

```ts
const { data } = useSwr(["user", sessionId])
// vs
const { mutate } = useSwr("user")
useEffect(() => mutate(), [mutate, sessionId])
```

The issues with the first variant are:

1. `data` becomes undefined for a moment when `sessionId` changes
2. it retains multiple composite keys for "old" users associated with old `sessionIds`
3. makes it harder to mutate globally, since the key is dependent on the current `sessionId`

The second variant has none of the above mentioned drawbacks, but its more boilerplace code

What is recommended by the team at vercel? @shuding

**Comment by shuding** [comment, marked as answer]
Yeah this is a great question! For this specific use case, we use the second variant because "user" is a global singleton. And we also have an effect to revalidate when:
- sessionId changes
- logging out

Seems like a special case though, but good to have a section or example in the documentation. I'll note this down!
